include/configoptions.h

```diff
@@ -0,0 +1,361 @@
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef SMARTREDIS_CONFIGOPTIONS_H
+#define SMARTREDIS_CONFIGOPTIONS_H
+
+#ifndef __cplusplus
+#error C users should include c_configoptions.h, not configoptions.h
+#endif
+
+#ifdef __cplusplus
+#include <stdlib.h>
+#include <string>
+#include <vector>
+#include <unordered_map>
+#include "srobject.h"
+#include "sr_enums.h"
+
+///@file
+
+namespace SmartRedis {
+
+/*!
+*   \brief Configuration source enumeration
+*/
+enum cfgSrc {
+    cs_file,    // Configuration data is coming from a JSON file
+    cs_envt,    // Configuration data is coming from environment variables
+    cs_blob,    // Configuration data is coming from a JSON blob string
+    cs_default  // Configuration data is coming from the default source
+} ;
+
+
+/*!
+*   \brief The ConfigOptions class bottlenecks all configuration
+*          options used in SmartRedis and allows them to be accessed
+*          from environment variables, from a JSON file, or from a
+*          JSON blob supplied as a text string.
+*/
+class ConfigOptions
+{
+    private:
+        /*!
+        *   \brief ConfigOptions constructor. Do not use! To instantiate
+        *          a ConfigOptions object, use one of the factory methods
+        *          below.
+        *   \param source The selected source for config data
+        *   \param string The string associated with the source
+        */
+        ConfigOptions(cfgSrc source, const std::string& string);
+
+    public:
+
+        /*!
+        *   \brief ConfigOptions copy constructor
+        *   \param cfgopts The ConfigOptions to copy
+        */
+        ConfigOptions(const ConfigOptions& cfgopts) = default;
+
+        /*!
+        *   \brief ConfigOptions copy assignment operator
+        *   \param cfgopts The ConfigOptions to copy and assign
+        */
+        ConfigOptions& operator=(const ConfigOptions& cfgopts) = default;
+
+        /*!
+        *   \brief ConfigOptions move constructor
+        *   \param cfgopts The ConfigOptions to move
+        */
+        ConfigOptions(ConfigOptions&& cfgopts) = default;
+
+        /*!
+        *   \brief ConfigOptions move assignment operator
+        *   \param cfgopts The ConfigOptions to move and assign
+        */
+        ConfigOptions& operator=(ConfigOptions&& cfgopts) = default;
+
+        /*!
+        *   \brief ConfigOptions destructor
+        */
+        virtual ~ConfigOptions() = default;
```

Comment:
If we aren't allowing configuration of defaults, this variable goes away and the point is moot.

With that said, my original implementation did have the object itself in the static space, not a pointer to it. This was causing some compilers to issue double deletes and making our CI/CD tests fail.  Since it's a single instance with lifetime equal to the duration of the containing library, the impact of this leak was minimal. Here, the small leak was the lesser evil